### PR TITLE
Update action after validate payment

### PR DESCRIPTION
**Pull request by dzungtran**

### CODE REVIEW

File: src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php [hunk 1]
```diff
@@ -114,4 +109,70 @@ private function prepareUserProfileData($customer)
...
+                break;
+
+            case 'wait':
+                $invoice->sendNotification('payment_waiting', $invoice->member, $validate['publicNote']);
+                break;
+
+            case 'fail':
+                $invoice->sendNotification('payment_failed', 0, $transaction);
+                break;
+        }
```
Comment (sevastyanovio) [inline src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php:227]: move to another method
try to keep code simple as possible

File: src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php [hunk 2]
```diff
@@ -114,4 +109,70 @@ private function prepareUserProfileData($customer)
...
+
+        //-----------------------------------------
+        // Is the invoice paid now?
+        //-----------------------------------------
+
+        $paid = $this->DB->buildAndFetch(array('select' => 'SUM( t_amount ) as paid', 'from' => 'nexus_transactions', 'where' => "t_status='okay' AND t_invoice={$invoice->id}"));
+        if (round(floatval($paid['paid']), 2) >= floatval($invoice->total)) {
+            $invoice->markPaid();
+        }
+
```
Comment (sevastyanovio) [inline src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php:200]: move to another method
try to keep code simple as possible

File: src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php [hunk 3]
```diff
@@ -114,4 +109,70 @@ private function prepareUserProfileData($customer)
...
+            't_extra' => serialize($extra),
+            't_gw_id' => $validate['gw_id'],
+            't_date' => time(),
+        );
+        if ($validate['amount']) {
+            $save['t_amount'] = $validate['amount'];
+        }
+        $transaction = array_merge($transaction, $save);
+        $this->DB->update('nexus_transactions', $save, "t_id={$transaction['t_id']}");
+
```
Comment (sevastyanovio) [inline src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php:170]: move to another method
try to keep code simple as possible

File: src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php [hunk 4]
```diff
@@ -114,4 +109,70 @@ private function prepareUserProfileData($customer)
...
+        // Log
+        if ($invoice->member) {
+            customer::load($invoice->member)->logAction('transaction', array('type' => 'paid', 'status' => $validate['status'], 'id' => $transaction['t_id'], 'invoice_id' => $invoice->id, 'title' => $invoice->title));
+        }
```
Comment (sevastyanovio) [inline src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php:184]: move to another method
try to keep code simple as possible

File: src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php [hunk 5]
```diff
@@ -114,4 +109,70 @@ private function prepareUserProfileData($customer)
...
+
+        // Sort our values
+        $extra = unserialize($transaction['t_extra']);
+        $extra['note'] = $validate['note'];
+        $extra['publicNote'] = $validate['publicNote'];
+        if (isset($validate['extra']) and is_array($validate['extra'])) {
+            $extra = array_merge($extra, $validate['extra']);
+        }
+
```
Comment (sevastyanovio) [inline src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php:145]: move to another method
try to keep code simple as possible

File: src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php [hunk 6]
```diff
@@ -114,4 +109,70 @@ private function prepareUserProfileData($customer)
...
     }
+
+    private function afterValidatePayment($validate, $invoice, $transaction)
+    {
+        // Did that make any sense?
+        if (!in_array($validate['status'], array('okay', 'hold', 'fail', 'wait'))) {
+            $validate['status'] = 'fail';
+            $validate['note'] = 'err_naughty_gateway';
+        }
+
```
Comment (sevastyanovio) [inline src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php:132]: move to another method
try to keep code simple as possible

File: src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php [hunk 7]
```diff
@@ -54,51 +54,46 @@ public function payScreen()
...
         $pingback = new Paymentwall_Pingback($_GET, $_SERVER['REMOTE_ADDR']);
-        //$result = array();
+
         if ($pingback->validate()) {
 
+            $invoice = new invoice($pingback->getProductId());
+            $transaction = $this->DB->buildAndFetch(array('select' => '*', 'from' => 'nexus_transactions', 'where' => "t_id={$pingback->getParameter('id')}"));
+
+            if(!$invoice || !$transaction){
+                die('Invoice or Transaction invalid!');
```
Comment (sevastyanovio) [inline src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php]: `Invoice or Transaction is invalid`

File: src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php (hunk 4 above)
Comment (dzungtran) [inline src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php:184]: do you want create a function for one line of codes?

``` php
customer::load($invoice->member)->logAction('transaction', array('type' => 'paid', 'status' => $validate['status'], 'id' => $transaction['t_id'], 'invoice_id' => $invoice->id, 'title' => $invoice->title));
```

I can refactor codes to more readable.

File: src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php [hunk 8]
```diff
@@ -112,6 +114,134 @@ private function prepareUserProfileData($customer)
...
+        if ($invoice->member) {
+            customer::load($invoice->member)->logAction('transaction', array(
+                'type' => 'paid',
+                'status' => $validate['status'],
+                'id' => $transaction['t_id'],
+                'invoice_id' => $invoice->id,
+                'title' => $invoice->title
+            ));
+        }
+
```
Comment (sevastyanovio) [inline src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php:138]: move it to external method

File: src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php [hunk 9]
```diff
@@ -112,6 +114,134 @@ private function prepareUserProfileData($customer)
...
+        if (!in_array(
+            $validate['status'],
+            array(
+                self::TRANSACTION_STATUS_OKAY,
+                self::TRANSACTION_STATUS_HOLD,
+                self::TRANSACTION_STATUS_FAIL,
+                self::TRANSACTION_STATUS_WAIT
+            ))
+        ) {
+            $validate['status'] = self::TRANSACTION_STATUS_FAIL;
```
Comment (sevastyanovio) [inline src/admin/applications_addon/ips/nexus/sources/gateways/paymentwall.php:129]: move it to external method, which return bool

**Pull request merged**